<a href="https://colab.research.google.com/github/Bhavyamadhuri/Bhavya_INFO5731_Fall2024/blob/main/INFO5731_Assignment_1_Spring_2024.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [1]:
# Your code here
import math

# Function for alculating the volume V of a hemisphere with radius r
def volume_of_hemisphere(r):
    return (2/3) * math.pi * r**3

# For the range of even numbers between 1 and 20:
for r in range(2, 21, 2):
    volume = volume_of_hemisphere(r)
    print(f"r {r} V {volume:.2f}")



r 2 V 16.76
r 4 V 134.04
r 6 V 452.39
r 8 V 1072.33
r 10 V 2094.40
r 12 V 3619.11
r 14 V 5747.02
r 16 V 8578.64
r 18 V 12214.51
r 20 V 16755.16


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [2]:
# Write code for each of the sub parts with proper comments.
#importing the 'random' module for generating random numbers using the 'random' function:
import random

# (a) This Generates a sequence of 12 random numbers within the range of 50 to 150
random_numbers = [random.randint(50, 150) for _ in range(12)] #assigned random numbers to the variable "random_numbers" using the "randint" function
print("Generated random numbers:", random_numbers) #printing the generated random numbers

# (b) Arrange the generated numbers in ascending order
sorted_ascending = sorted(random_numbers) #Sorting the generated numbers in ascending order using the "sorted" function
print("\nNumbers in ascending order:", sorted_ascending) #printing the numbers in ascending order

# (c) Arrange the numbers in descending order
sorted_descending = sorted(random_numbers, reverse=True) #Sorting the generated numbers in descending order using the "sorted" function, the parameter "reverse" is given "True" so that the numbers will be reversed, after soring them in ascending order.
print("\nNumbers in descending order:", sorted_descending) #printing the numbers in descending order

# (d) Compute and output the average of the entire group of generated numbers
average_total = sum(random_numbers) / len(random_numbers) #First calculating the sum of numbers and diving them with the length gives us the average
print("\nAverage of the entire group:", round(average_total, 2)) #printing the average of numbers

# (e) Calculate and display the average of the first 6 numbers in both ascending and descending lists
average_first_6_ascending = sum(sorted_ascending[:6]) / 6 #this gives us the average of first six numbers is ascending order
average_first_6_descending = sum(sorted_descending[:6]) / 6 #this gives us the average of first six numbers is descending order
print("\nAverage of the first 6 numbers (ascending):", round(average_first_6_ascending, 2)) #printing the average of first six numbers in ascending order
print("Average of the first 6 numbers (descending):", round(average_first_6_descending, 2)) #printing the average of first six numbers in descending order



Generated random numbers: [133, 75, 118, 145, 125, 149, 74, 106, 55, 118, 67, 60]

Numbers in ascending order: [55, 60, 67, 74, 75, 106, 118, 118, 125, 133, 145, 149]

Numbers in descending order: [149, 145, 133, 125, 118, 118, 106, 75, 74, 67, 60, 55]

Average of the entire group: 102.08

Average of the first 6 numbers (ascending): 72.83
Average of the first 6 numbers (descending): 131.33


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [3]:
# Your code here

# Your code here
import string #This line of code imports the string module

# (a) Prompt the user to input two sentences
sentence1 = input("Enter the first sentence: ") #This line inputs the first sentence
sentence2 = input("Enter the second sentence: ") #This line inputs the second sentence

# (b) Transform each sentence into a list of words, including punctuation as separate elements
sentence1_list = sentence1.split()# The split function seperates the elemenents in a sentence
sentence2_list = sentence2.split()

# Combine both lists into a single list to show as output
combined_sentence_list = sentence1_list + sentence2_list #Adding both the sentences to show as a single sentence.
print("\nCombined word list from both sentences:")
print(combined_sentence_list) #this line prints the combined sentences

# (c) Compare the two sentences and find common words
common_words = set(sentence1_list) & set(sentence2_list) #This finds the common words using the '&' and gives the results in the form of sets
print("\nCommon words between both sentences:")
print(list(common_words)) #prints the common words as a list

# (d) Remove any punctuation from both sentences
def remove_punctuation(sentence):  #defining a function for future use for removing the punctuation
    return sentence.translate(str.maketrans("", "", string.punctuation)) #maketrans uses a translation table and translate method removes all the punctuations. The first argument means no characters are being mapped to anything else

sentence1_sanitized = remove_punctuation(sentence1) #removes punctuation from sentence 1
sentence2_sanitized = remove_punctuation(sentence2)  #removes punctuation from sentence 1
print("\nSanitized Sentence 1 (punctuation removed):")
print(sentence1_sanitized) #prints the sanitized sentence 1
print("Sanitized Sentence 2 (punctuation removed):")
print(sentence2_sanitized) #prints the sanitized sentence 2

# (e) Count the number of vowels in each sentence
def count_vowels(sentence):  #defining a function for counting the number of vowels in each sentence
    vowels = "aeiouAEIOU"
    vowel_count = {vowel: 0 for vowel in vowels.lower()} #using for loop and adding number of vowels in the sentence
    for char in sentence:
        if char.lower() in vowels:
            vowel_count[char.lower()] += 1
    return vowel_count   #returns the number of vowels to the function count_vowels as output parameter

vowels_sentence1 = count_vowels(sentence1_sanitized) #using the function defined to the count the number of vowels in sentence1_sanitized
vowels_sentence2 = count_vowels(sentence2_sanitized) #using the function defined to the count the number of vowels in sentence2_sanitized

# Print vowel counts for both sentences
print("\nVowel count for Sentence 1:")
for vowel, count in vowels_sentence1.items():
    print(f"{vowel}: {count}") #prints the vowels and their corresponding counts in Sentence 1 with the help of formatting

print("\nVowel count for Sentence 2:")
for vowel, count in vowels_sentence2.items():
    print(f"{vowel}: {count}") #prints the vowels and their corresponding counts in Sentence 2 with the help of formatting

Enter the first sentence: My name Bhavya Madhuri!!
Enter the second sentence: my last name is Devarakonda

Combined word list from both sentences:
['My', 'name', 'Bhavya', 'Madhuri!!', 'my', 'last', 'name', 'is', 'Devarakonda']

Common words between both sentences:
['name']

Sanitized Sentence 1 (punctuation removed):
My name Bhavya Madhuri
Sanitized Sentence 2 (punctuation removed):
my last name is Devarakonda

Vowel count for Sentence 1:
a: 4
e: 1
i: 1
o: 0
u: 1

Vowel count for Sentence 2:
a: 5
e: 2
i: 1
o: 1
u: 0


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [5]:
# Your code here
import math #this line of code imports the math module

#Defining a function to evaluate a given expression
def evaluate_expression(expression):
    try:      #using the expcetional handling techniques to get different results for different error types
        # (a) and (b): Evaluate the mathematical expression
        result = eval(expression)
        return result
    except (SyntaxError, NameError): # this print the error message when an invalid expression is entered like a syntax error
        return "Error: Invalid expression. Please check your input."
    except ZeroDivisionError:   #this prints the below line when any given value is divided by zero
        return "Error: Division by zero is not allowed."
    except Exception as e:
        return f"Error: {str(e)}"

#Defining a function for the calculator
def run_calculator():
    print("Welcome to the Command-line Calculator!")
    print("Supported operations: +, -, , /, * (exponentiation), % (modulus), sqrt()")
    print("You can use parentheses for grouping and type 'exit' to quit.")

#Using while loop as we can use the calculator as long as we want by continually running the program and we can stop by terminating the program once done by typing 'exit'
    while True:
        expression = input("\nEnter a mathematical expression (or type 'exit' to quit): ").strip()

        # (e) Command to exit the calculator
        if expression.lower() == 'exit':
            print("Exiting the calculator. Goodbye!")
            break

        # (c) Support for square root function
        expression = expression.replace('sqrt', 'math.sqrt')

        # (a), (b), (c) Evaluate the expression
        result = evaluate_expression(expression)

        # Display the result
        print("Result:", result)

# This is the main function from where the program will start running the calculatorand  and cannot be imported to another file
if __name__ == "__main__":
    run_calculator()

Welcome to the Command-line Calculator!
Supported operations: +, -, , /, * (exponentiation), % (modulus), sqrt()
You can use parentheses for grouping and type 'exit' to quit.

Enter a mathematical expression (or type 'exit' to quit): 2+3
Result: 5

Enter a mathematical expression (or type 'exit' to quit): 2*6
Result: 12

Enter a mathematical expression (or type 'exit' to quit): exit
Exiting the calculator. Goodbye!


# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [10]:
 # Your code here

import pandas as pd

# (A)
def calculate_occupied_housing_difference(file_path):

    df = pd.read_csv(file_path)

    occupied_df = df[df['title_field'] == 'Occupied Housing Units']

    occupied_df = occupied_df.sort_values(by='year')

    occupied_df['Difference'] = occupied_df['value'].diff()

    occupied_df = occupied_df.dropna(subset=['Difference'])

    sorted_diff_df = occupied_df[['year', 'Difference']].sort_values(by='Difference')

    print("Year-to-Year Differences in Occupied Housing Units (sorted in ascending order):")
    print(sorted_diff_df)

import pandas as pd
from scipy.stats import hmean

# Parts (b,c,d)
def housing_stats(file_path):

    df = pd.read_csv(file_path)

    # Part (b)
    total_df = df[df['title_field'] == 'Total Housing Units']
    occupied_df = df[df['title_field'] == 'Occupied Housing Units']
    vacant_df = df[df['title_field'] == 'Vacant Housing Units']

    percentages = []
    for year in total_df['year'].unique():
        total_units = total_df[total_df['year'] == year]['value'].values[0]
        occupied_units = occupied_df[occupied_df['year'] == year]['value'].values[0]
        vacant_units = vacant_df[vacant_df['year'] == year]['value'].values[0]

        occupied_percent = (occupied_units / total_units) * 100
        vacant_percent = (vacant_units / total_units) * 100

        percentages.append((year, vacant_percent, occupied_percent))

    print("Year Vacant Housing Units Occupied Housing Units")
    for year, vacant, occupied in percentages:
        print(f"{year}  {vacant:.2f}%   {occupied:.2f}%")

    # Part (c)
    max_vacant_year = vacant_df.loc[vacant_df['value'].idxmax()]['year']
    max_vacant_value = vacant_df['value'].max()

    max_occupied_year = occupied_df.loc[occupied_df['value'].idxmax()]['year']
    max_occupied_value = occupied_df['value'].max()

    print("\n                              Year     Value")
    print(f"Vacant Housing Units      {max_vacant_year}   {max_vacant_value}")
    print(f"Occupied Housing Units    {max_occupied_year}   {max_occupied_value}")

    # Part (d)
    total_units = total_df['value'].values
    harmonic_mean = hmean(total_units)

    print(f"\nHarmonic Mean of Total Housing Units: {harmonic_mean:.2f}")





# File path to the CSV file
file_path = '/Users/meghamshreddy/Desktop/Assignment1_denton_housing.csv'  # Replace with your actual file path



# Call the function
calculate_occupied_housing_difference(file_path)

# Call the function
housing_stats(file_path)


FileNotFoundError: [Errno 2] No such file or directory: '/Users/meghamshreddy/Desktop/Assignment1_denton_housing.csv'

# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
# Write your response below